# Wrangling Genomics

Lesson for quality control and wrangling genomics data. This repository is maintained by
[Josh Herr][josh_herr], [Ming Tang][ming_tang], [Lex Nederbragt][lex_nederbragt], and
[Fotis Psomopoulos][fotis_psomopoulos].

Amazon public AMI for this tutorial is "dataCgen-qc".

## Background

Wrangling genomics trains novice learners on a variant calling workflow. Participants will learn how
to evaluate sequence quality and what to do if it is not good. We will then cover aligning reads to
a genome, and calling variants, as well as discussing different file formats. Results will be
visualized. Finally, we will cover how to automate the process by building a shell script.

This lesson is part of the [Data Carpentry][data_carpentry] [Genomics Workshop][genomics_workshop].

## Contribution

- Make a suggestion or correct an error by [raising an Issue][genomics_issues].

## Code of Conduct

All participants should agree to abide by the [Data Carpentry Code of Conduct][dc_coc].

## Authors

Wrangling genomics is authored and maintained by the [community][genomics_members].

## Citation

Please cite as:

Wrangling Genomics. June 2017. http://datacarpentry.github.io/wrangling-genomics/.

[data_carpentry]: http://www.datacarpentry.org
[dc_coc]: http://www.datacarpentry.org/code-of-conduct
[fotis_psomopoulos]: https://github.com/fpsom
[genomics_issues]: https://github.com/datacarpentry/wrangling-genomics/issues
[genomics_members]: https://github.com/datacarpentry/wrangling-genomics/network/members
[genomics_workshop]: http://www.datacarpentry.org/genomics-workshop
[josh_herr]: https://github.com/jrherr
[lex_nederbragt]: https://github.com/lexnederbragt
[ming_tang]: https://github.com/crazyhottommy
